File: backend/app/api/routes/items.py
from fastapi import APIRouter

from app.api.deps import SessionDep
from app.crud.item import create_item, get_items
from app.models.item import Item
from app.schemas.item import ItemCreate, ItemRead

router = APIRouter()


@router.post("/items", response_model=ItemRead)  # response_model → the public output shape
def add_item(item_in: ItemCreate, session: SessionDep) -> Item:
    return create_item(session, item_in)


@router.get("/items", response_model=list[ItemRead])
def list_items(session: SessionDep) -> list[Item]:
    return get_items(session)

File: frontend/src/api/items.ts
import { z } from 'zod'

import axiosInstance from './axiosInstance'
import { itemSchema, type Item, type ItemCreate } from '../schemas/item'

export async function getItems(): Promise<Item[]> {
    const response = await axiosInstance.get('/items')
    return z.array(itemSchema).parse(response.data)  // validate the response at runtime
}

export async function createItem(data: ItemCreate): Promise<Item> {
    const response = await axiosInstance.post('/items', data)
    return itemSchema.parse(response.data)
}
